### ISSUE TITLE
Add rendering distance 

### ISSUE THREAD

**Issue by raspofabs** [raspofabs/OodleCraft#20]
With larger amounts of land, and lower machine specs, rendering will be a bottleneck. Add some form of distance based rendering limit.

**Comment by Dala0** [comment]
On Window 7 it already appears to have a distance based rendering limit. If you get far enough away from blocks, they will not render anymore.

It would be useful if this limit was user configurable eaiserly, for example by changing a number in a settings file or via the keyboard.

**Comment by raspofabs** [comment]
That's just far clipping. The editor is still rendering all the cubes, just they are not visible because they are clipped off the back of the Z buffer.

**Comment by Dala0** [comment]
I see. The Z buffer limit is set where? In OpenGL?

**Comment by raspofabs** [comment]
yes, and whoever does the render distance should also adjust the farclip as well, but not based on render distance, but furthest render distance.

**Comment by Dala0** [comment]
Can we set the value of the Z buffer limit?

If so, would that be a good way to control rendering?

**Comment by Dala0** [comment]
I am having a look in the source trying to find where these things should be done. The code in drawing.py looks like the right place, what with the OpenGL commands.

**Comment by Dala0** [comment]
I had a go at changing the commands in drawing.py. I found how to change the background colour and viewing angle.

There is a lot of code in here that might relate to viewing distance. Any chance of a comment along the lines of 'Here is the code that needs to be changed'?

**Comment by raspofabs** [comment]
Drawing. Py had lots of drawing related stuff, but the real render range comes from what land is loaded. The Z depth is controlled through the protection matrix. Making this really small won't make things better for low spec machines.

**Comment by Dala0** [comment]
So we need some code that loads certain blocks and removes other blocks based on distance from the camera.

Do you think it should be better if individual blocks where swapped in and out of memory or would it work better doing groups of blocks in one go?

**Comment by Dala0** [comment]
I have been thinking about this. If we go for a system that loads blocks one at a time, then I think we might run into issues with speed of access to storage media.

I am thinking about the difference between two possible systems.

System one. This system loads blocks in a one unit cubed space when needed.

System two. This system loads blocks in a thirty two unit cubed space when needed.

System two could end up needing around 32767 less accesses to the storage media than system one to load one thirty two unit cubed area.

**Comment by Dala0** [comment]
I am trying to work out how the data is stored.

The data segment below: -

s(ivec
Vec3
p307999
(dp308000
g287841
F426.4156110857885
sg287842
F250.64749450713853
sg287843
F118.76602184784629
sbF-2.9709999999996248
tp308001
sI30
((dp308002
g287363
I1

looks like it might refer to a land, with information about the coordinates and orientation in absolute terms.

The data segment below: -

tp6
I17
s(I136
I405
I110

looks like it might refer to a block within a land, with information about it in terms relative to the origin point of the land.

Below is some speculation as to the what this data means: -

tp6 = block designation
I17 = texture
s(I136 = distance from the origin on the x axis
I405 = distance from the origin on the y axis
I110 = distance from the origin on the z axis

**Comment by Dala0** [comment]
I am thinking that it might be a good idea to split up the save file into various files based on position in space.

If we keep the current system of one file for all the data, then we are doing to have to read the entire file every time we want to read block position data. This could start to be a problem at some point, as size of the data increases.

One possible system would be to split up the space into cubes of say 4096 units. New files would be created when blocks where spawned in a cube without a related file. This system has a possible problem. What do we do about lands with origins in another cube? Maybe if a land has data in a cube where its origin is not we could leave a bit of data saying which file the land data is in.
Another way around this problem might be to clone the land position data into every save cube file it has blocks in and store the blocks for that land in the local cube save file even though the origin of the land is elsewhere.

**Comment by Dala0** [comment]
What do you think about putting the data in a database such as MySQL? We could have one table for land data. This table would include the current dimensions of the land in terms of outlying blocks. If there  are two blocks in the land with the cords 0-0-0 and 3-0-1, then store the cords of the extent of the land as 0-0-0, 3-0-0, 0-1-0, 3-1-0, 0-1-0, 3-1-0, 0-1-2, and 3-1-2. These cords can then be read to find out if the land bisects a given volume.

A second table would store the block data. In this table, one of the columns would store a reference number for the land the block was in.

Strengths of this system vs files.

1. Possible speed benefits, as it would be possible to only load the data needed for a given task using an SQL statement.
2. MySQL tools such as MySQL Workbench become useable with the data.
3. MySQL can be configured to cache data in RAM. This may out-perform reading data from a storage device.

Weaknesses of this system vs files.

1. Requires database software installation.
2. Requires a more complex rewrite of the current code to work.
3. Math working out if an added block changes the dimensions of a land would need to be executed every time a block was spawned and if the dimensions where different then the land table would need to be updated. Similar code would be needed on removing blocks.

**Comment by Dala0** [comment]
I have been thinking more about the possibility of putting the data into MySQL tables. There is a good chance we can come up with a better solution than the one I outlined above for referencing the extent of blocks in lands.

One of the best ways might be to do similar calculations using SQL either every time they are needed or to use SQL to update the land table with information about the extent of blocks in it periodically.

It would be useful to know a maximum number of lands that system should be functional with, as that may affect the design of the system.

**Comment by Dala0** [comment]
I have thought of another possibly useful way to implement the system in MySQL. We could have a third table that basically acts like the cubes in my file system based system design. The table would have a large number of columns. Most of the columns would store reference numbers of lands that bisect the cube's volume.

Each cube would have it's own row.
